### ISSUE TITLE
support custom names on resource imports

### ISSUE THREAD

**Issue by mwalle** [robotframework/robotframework#2104]
This issue depends on #1999.

Adding support for custom names will solve the problem with duplicate resource filenames. See here fore more information: https://groups.google.com/d/msg/robotframework-users/0Cb215YSnG4/mWpGHi5dwUwJ

Syntax:

```
Import Resource  path/to/resource.txt  WITH NAME  my_resource
Resource  path/to/resource.txt  WITH NAME  my_resource
```

**Comment by pekkaklarck** [comment]
This is good idea but I don't consider it too high priority. If duplicate resource names are a problem, it is typically fairly easy to rename them. With libraries custom names are more important because you may want to use same library with different parameters.

**Comment by hthiery** [comment]
We have to support user keyword files for different type of devices. As long you write a test only for one device the problem does not occure, because you only import one resource. But when you start writing tests with more than one test in the setup you will hit this issue.

Example:

Resourcefile: supportTypeDevice1/control.txt

    *** Keywords ***
    Reset Device
        Do something to reset device type 1


Resourcefile: supportTypeDevice2/control.txt

    *** Keywords ***
    Reset Device
        Do something to reset device type 2

Testcase-file.txt

    Resource   ${RESOURCES}/supportTypeDevice1/control.txt
    Resource   ${RESOURCES}/supportTypeDevice2/control.txt

    *** Test Cases ***
    Test1
        control.Reset Device     # want to reset device type 1
        control.Reset Device     # want to reset device type 2

With the changes we could do the following;

Testcase-file.txt

    Resource   ${RESOURCES}/supportTypeDevice1/control.txt  WITH NAME  control_device_1
    Resource   ${RESOURCES}/supportTypeDevice2/control.txt  WITH NAME  control_device_2

    *** Test Cases ***
    Test1
        control_device_1.Reset Device
        control_device_2.Reset Device

**Comment by dkindt** [comment]
> This is good idea but I don't consider it too high priority. If duplicate resource names are a problem, it is typically fairly easy to rename them. With libraries custom names are more important because you may want to use same library with different parameters.

Are we still feeling this way almost 4 years later? This would be a _very_ helpful feature and I'm sure there would be plenty in the community to help? Where would I get started to even look into this?

**Comment by pekkaklarck** [comment]
Yeah, I still see this as a nice to have but not too high priority feature. That means I won't implement it myself in the foreseeable future, but I would be super happy to review and merge PR if someone else finds this more important.

I don't remember what code to start modifying, but can look at that later if needed. There are two problems implementing this, though:
- We are currently rewriting parser, and parsing related code needs to wait until the new parser is ready to be merged to master.
- Code related to imports is old and not that good. Larger rewrite would be nice but it should be separately.

**Comment by mattivityroom** [comment]
This feature are not implemented yet ?

**Comment by drchopper** [comment]
> This feature does not implemented yet ?

Unfortunately, no.

**Comment by pekkaklarck** [comment]
This was recently proposed again as #3841 so there's clearly some interest. I still don't consider this too high priority issue because typically you can simply name different resource files differently. That said, I'm not against adding this feature (I'd close this issue if I was) especially if someone is interested to help with it.

**Comment by logix-md** [comment]
> This was recently proposed again as #3841 so there's clearly some interest. I still don't consider this too high priority issue because typically you can simply name different resource files differently. That said, I'm not against adding this feature (I'd close this issue if I was) especially if someone is interested to help with it.

Unfortunately, this is not always possible to use different resource file name without redesign whole project. For example I've faced with this problem when I need to do upgrade testing... We have several folders with with  resources for several versions of tested product. We need to use resources with the same name  which located in different folders, but now it's not possible .

`
*** Settings ***
Resource   ${RESOURCES}/CommonKW.robot
Resource   ${RESOURCES}/v1.0/control.robot
Resource   ${RESOURCES}/v2.0/control.robot

*** Test Cases ***
Test1
    UIControl.Configure Product        # Want to Configure Product of version 1
    CommonKW,Upgrade Product     2.0   
    UIControl.Configure Product        # Want  to Configure Product of version 2
`

Also, 'BuiltIn.Import Resource' use the test suite scope similarly when importing them in the Setting table using the Resource settings and there is not way to  'reimport' resource  from other folder without 'Multiple keywords with name...' error.

**Comment by nathanhannig** [comment]
This would be a very nice feature

**Comment by linyy1106** [comment]
Wondering when this nice feature will be available?

**Comment by crayfishuk** [comment]
We are using Resource files as Page Model resources, and to organise them we have a folder structure that mimics our MVC layout for the Controllers/Views.  We often have multiple views with the same name (e.g. edit.resource) -- but in different folders (as they correspond with different Controllers (e.g. clients/edit.resource and articles/edit.resource). Obviously we could rename the resource files to flatten into one folder  -- but this then becomes very large and cumbersome.   

When closely coupled tests move between Controllers, they often conflict.  It would be useful to have the ability to usee:
```
Resource    ../../pages/clients/edit.resource    WITH NAME     ClientsEdit
Resource    ../../pages/articles/edit.resource    WITH NAME    ArticleEdit
:  
:
     ClientsEdit.Enter Into Notes     some text
     ClientsEdit.Click Submit
:
:    
     ArticleEdit.Enter Into Notes     some article text
     ArticleEdit.Click Submit
```

It feels a consistent usage with the `Library ... WITH NAME ....` and I was surprised it wasn't there...

**Comment by pekkaklarck** [comment]
The main reason we support aliases with libraries is to allow importing same lib with different arguments. That doesn't apply for resource files, but I can see just renaming libs and resources at import time is useful as well.

This probably wouldn't be overly complicated to implement, but the whole import system should get a rewrite and I'm not too interested to add new features to it until that happens. That said, i someone else is interested to look at this sooner, I'm happy to help.

If/when this is implemented, the marker to use should be `AS`, not `WITH NAME`. See #4369 for details.

**Comment by jura84** [comment]
Nice feature request and I would be more than happy to see this in Robot.

Any chance to get priority for this one? In our project we have lot of resource files where we categorize certain keyword in specific file.

When using AS/WITH NAME would make it easier to see where callerd keyword is located.

**Comment by gblack72** [comment]
I'm trying RF two weeks for attempt to test our web application. Everyone who make bigger then small app uses namespaces for imports. Than goes to the hell with this discomfort in RF.
Maybe I have other suggestion. Other solution could be to have possible qualify keyword not only by resource name but with folders too.
So for the end, how many developers should notice that they are missing this feature to increase the priority, I know a few of them? :-)

**Comment by pekkaklarck** [comment]
This issue is about renaming resource files on import like
```
Resource       X.resource    AS    Y
```
which would basically would work similarly as
```python
import x as y
```
in Python. I can see that it would be convenient, but because you can also just rename that `X.resource` to `Y.resource`, I don't consider this super important. We do support this with libraries, but with them the main use case is importing same library with different parameters like
```
Library         X    arg=1    AS    X1
Library         X    arg=2    AS    X2
```
Because this is pretty low priority issue and implementation isn't trivial, this isn't currently targeted to any Robot version. If someone considers this important and is interested to provide a pull request, I'd be happy to help.

**Comment by gblack72** [comment]
IMO is this not primary about "renaming resource files on import". That is proposed solution becouse of resource names conflicts.
You have been defending to realize it for years. So then I'm proposing other solution. Meybe could be easier to implement it.
But ok. That we can figure out.
folder1/folder2/folder3/**section1**/folder4/_left-menu.resources_
folder1/folder2/folder3/**section2**/folder4/_left-menu.resources_
...
rename to 
folder1/folder2/folder3/**section1**/folder4/_folder1_folder2_folder3_section1_folder4_folleft-menu.resources_
folder1/folder2/folder3/**section2**/folder4/_folder1_folder2_folder3_section2_folder4_folleft-menu.resources_
...
But everyone has to see it's little bit uncomfortable. :-)

**Comment by crayfishuk** [comment]
@gblack72 - indeed - that is fine as a work-around; but a complete nightmare for a) readability, b) maintainability, c) keyboard life-time!

Are there any plans for some kind of 'Namespace' feature in RF?

**Comment by Redysz** [comment]
I have different situation. I have two different resources, let's say `power_supply.resource` and `board.resource`. I would like to not care if some keywords are repeated. Now I have `Is On` in both resources so when importing both I can't say which one will be used (if any). With aliases it would be:
```
Resource    power_supply.resource  AS PS
Resource    board.resource  AS BOARD
[...]
    ${Power Supply Is On}  PS.Is On
    ${Board Is On}  BOARD.Is On
```
Any work-around for this without changing keywords names?

**Comment by d-biehl** [comment]
push!!!
